src/java.net.http/share/classes/jdk/internal/net/http/Http2TerminationCause.java

```diff
@@ -0,0 +1,277 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.net.http;
+
+
+import java.io.IOException;
+import java.net.ProtocolException;
+import java.util.Objects;
+
+import jdk.internal.net.http.frame.ErrorFrame;
+
+/**
+ * Termination cause for a {@linkplain Http2Connection HTTP/2 connection}
+ */
+public abstract sealed class Http2TerminationCause {
+    private String logMsg;
+    private String peerVisibleReason;
+    private final int closeCode;
+    private final Throwable originalCause;
+    private final IOException reportedCause;
+
+    private Http2TerminationCause(final int closeCode, final Throwable closeCause) {
+        this.closeCode = closeCode;
+        this.originalCause = closeCause;
+        if (closeCause != null) {
+            this.logMsg = closeCause.toString();
+        }
+        this.reportedCause = toReportedCause(this.originalCause, this.logMsg);
+    }
+
+    private Http2TerminationCause(final int closeCode, final String loggedAs) {
+        this.closeCode = closeCode;
+        this.originalCause = null;
+        this.logMsg = loggedAs;
+        this.reportedCause = toReportedCause(null, this.logMsg);
+    }
+
+    /**
+     * Returns the error code (specified for HTTP/2 ErrorFrame) that caused the
+     * connection termination.
+     */
+    public final int getCloseCode() {
+        return this.closeCode;
+    }
+
+    /**
+     * Returns the IOException that is considered the cause of the connection termination.
+     * Even a normal {@linkplain #isErroneousClose() non-erroneous} termination will have
+     * a IOException associated with it, so this method will always return a non-null instance.
+     */
+    public final IOException getCloseCause() {
+        return this.reportedCause;
+    }
+
+    /**
+     * Returns {@code true} if the connection was terminated due to some exception. {@code false}
+     * otherwise.
+     * A normal connection termination (for example, the connection idle timing out locally)
+     * is not considered as an erroneous termination and this method returns {@code false} for
+     * such cases.
+     */
+    public abstract boolean isErroneousClose();
```

Comment:
nit: can we use a different word here? "Erroneous close" feels vague here; would "is(Non)Graceful", "isAbrupt" or "hasErrorCode" capture the intent?

Reply:
That that `erroneous close` has been used in several other contexts; in code, in comments, etc. If this gets updated, I'd appreciate other relevant occurrences get updated too.

Reply:
I updated the PR to use `isAbnormalClose()`. I can rename it further if necessary.